File: Classes/Extraction.vb
Imports System.IO
Public Class Extraction
    ''' <summary>
    ''' Checks if the file we want to access is locked or not.
    ''' A lock means, another process might be still writing the file, e.g. the scanner
    ''' </summary>
    ''' <param name="filePath"></param>
    ''' <returns></returns>
    Public Shared Function IsFileLocked(ByVal filePath As String) As Boolean
        Dim stream As FileStream = Nothing
        Try
            stream = File.Open(filePath, FileMode.Open, FileAccess.Read, FileShare.None)
        Catch ex As Exception
            Return True
        Finally
            If Not stream Is Nothing Then
                stream.Close()
            End If
        End Try
        Return False
    End Function

    Public Shared Sub Extraction(ByRef x As CompressedFile)
        If Not x.Processing Then
            x.Processing = True
        Else
            Exit Sub
        End If

        'Check if file is locked
        If IsFileLocked(x.FileInfo.FullName) Then
            ExitExtraction(x)
            Exit Sub
        End If
        '# ### #

        Dim sWatch As Stopwatch = New Stopwatch
        sWatch.Start()

        Dim newFolder As String = Path.Combine(ServiceConfig.operatingPath, Path.GetFileNameWithoutExtension(x.FileInfo.Name))
        Try
            If Not Directory.Exists(newFolder) Then
                Directory.CreateDirectory(newFolder)
            Else
                Log.Warning("| [~] Folder already exists, extraction aborted")
                ExitExtraction(x)
                Exit Sub
            End If
        Catch ex As Exception
            Log.Error(ex, "| [" & SpecialChars.GetChar(SpecialChars.Chars.Crossmark) & "] Error 3000: ")
            Exit Sub
        End Try

        Log.Information("| [+] Extraction started for """ & x.FileInfo.Name & """")
        Try
            Using archiveFile As SevenZipExtractor.ArchiveFile = New SevenZipExtractor.ArchiveFile(x.FileInfo.FullName)
                'Debug.Print(archiveFile.Entries.Count)
                For Each entry In archiveFile.Entries
                    If entry.IsFolder Then
                        Directory.CreateDirectory(Path.Combine(newFolder, entry.FileName))
                        x.EntriesDirectory += 1
                    Else
                        entry.Extract(Path.Combine(newFolder, entry.FileName))
                        x.EntriesFile += 1
                    End If
                Next
            End Using

            'tar.gz Exception (Addtional Extraction of TAR)
            If x.FileInfo.Name.EndsWith(".tar.gz") Then
                Dim tarName As String = Directory.GetFiles(newFolder).FirstOrDefault
                Dim newerFolder As String = Path.Combine(ServiceConfig.operatingPath, x.FileInfo.Name.Replace(".tar.gz", ""))
                If Not Directory.Exists(newerFolder) Then
                    Directory.CreateDirectory(newerFolder)
                End If

                Using archiveFile As SevenZipExtractor.ArchiveFile = New SevenZipExtractor.ArchiveFile(tarName)
                    'Debug.Print(archiveFile.Entries.Count)
                    For Each entry In archiveFile.Entries
                        If entry.IsFolder Then
                            Directory.CreateDirectory(Path.Combine(newerFolder, entry.FileName))
                            x.EntriesDirectory += 1
                        Else
                            entry.Extract(Path.Combine(newerFolder, entry.FileName))
                            x.EntriesFile += 1
                        End If
                    Next
                End Using

                If Directory.Exists(newerFolder) Then
                    Directory.Delete(newerFolder & ".tar", True)
                End If
            End If
            '# ### #


            sWatch.Stop()

            Log.Information("| [" & SpecialChars.GetChar(SpecialChars.Chars.Checkmark) & "] Extraction successful for """ & x.FileInfo.Name & """ in " & sWatch.Elapsed.TotalSeconds.ToString("0.##") & " Seconds")
            x.OperationTime = sWatch.Elapsed.TotalSeconds
        Catch ex As Exception
            Log.Error(ex, "| [" & SpecialChars.GetChar(SpecialChars.Chars.Crossmark) & "] Error 3001: ")
            ExitExtraction(x)
            Exit Sub
        End Try

        ExitExtraction(x, True)
    End Sub

    Private Shared Sub ExitExtraction(ByRef x As CompressedFile, ByVal Optional success As Boolean = False)
        With x
            .Processing = False
            .WasSuccessful = success
        End With
    End Sub
End Class

File: nxn-AutoExtractor/Classes/Extraction.cs
using nxn_AutoExtractor.Models;
using Serilog;
using System;
using System.Diagnostics;
using System.IO;
using System.Linq;
using neXn.Lib5.SpecialCharacters;
using neXn.Lib5.Files;
using SevenZipExtractor;

namespace nxn_AutoExtractor.Classes
{
    public class Extraction
    {
        internal CompressedFile compressedFile;
        internal string operatingPath;

        #region Constructor
        public Extraction(CompressedFile compressedFile, string operatingPath = null)
        {
            this.compressedFile = compressedFile;
            if (operatingPath != null)
            {
                this.operatingPath = operatingPath;
            }
            else
            {
                this.operatingPath = this.compressedFile.FileInfo.DirectoryName;
            }
        }
        #endregion

        public void Extract()
        {
            if (this.compressedFile.Processing)
            {
                return;
            }

            this.compressedFile.Processing = true;

            // Check if file is locked
            if (Core.IsFileLocked(this.compressedFile.FileInfo.FullName))
            {
                ExitExtraction();
                return;
            }
            // # ### #

            var sWatch = new Stopwatch();
            sWatch.Start();
            string newFolder = Path.Combine(this.operatingPath, Path.GetFileNameWithoutExtension(this.compressedFile.FileInfo.Name));
            try
            {
                if (!Directory.Exists(newFolder))
                {
                    Directory.CreateDirectory(newFolder);
                }
                else
                {
                    Log.Warning("| [~] Folder already exists, extraction aborted");
                    ExitExtraction();
                    return;
                }
            }
            catch (Exception ex)
            {
                Log.Error(ex, $"| [{Chars.Crossmark}] Error 3000: ");
                return;
            }

            Log.Information($"| [+] Extraction started for \"{this.compressedFile.FileInfo.Name}\"");
            try
            {
                using (var archiveFile = new ArchiveFile(this.compressedFile.FileInfo.FullName))
                {
                    foreach (var entry in archiveFile.Entries)
                    {
                        if (entry.IsFolder)
                        {
                            Directory.CreateDirectory(Path.Combine(newFolder, entry.FileName));
                            this.compressedFile.EntriesDirectory++;
                        }
                        else
                        {
                            entry.Extract(Path.Combine(newFolder, entry.FileName));
                            this.compressedFile.EntriesFile++;
                        }
                    }
                }

                // tar.gz Exception (Addtional Extraction of TAR)
                if (this.compressedFile.FileInfo.Name.EndsWith(".tar.gz"))
                {
                    string tarName = Directory.GetFiles(newFolder).FirstOrDefault();
                    string newerFolder = Path.Combine(this.operatingPath, this.compressedFile.FileInfo.Name.Replace(".tar.gz", ""));
                    if (!Directory.Exists(newerFolder))
                    {
                        Directory.CreateDirectory(newerFolder);
                    }
                    
                    using (var archiveFile = new ArchiveFile(tarName))
                    {
                        foreach (var entry in archiveFile.Entries)
                        {
                            if (entry.IsFolder)
                            {
                                Directory.CreateDirectory(Path.Combine(newerFolder, entry.FileName));
                                this.compressedFile.EntriesDirectory++;
                            }
                            else
                            {
                                entry.Extract(Path.Combine(newerFolder, entry.FileName));
                                this.compressedFile.EntriesFile++;
                            }
                        }
                    }

                    if (Directory.Exists(newerFolder))
                    {
                        Directory.Delete(newerFolder + ".tar", true);
                    }
                }
                // # ### #


                sWatch.Stop();
                Log.Information($"| [{Chars.Checkmark}] Extraction successful for \"" + this.compressedFile.FileInfo.Name + "\" in " + sWatch.Elapsed.TotalSeconds.ToString("0.##") + " Seconds");
                this.compressedFile.OperationTime = sWatch.Elapsed.TotalSeconds;
            }
            catch (Exception ex)
            {
                Log.Error(ex, $"| [{Chars.Crossmark}] Error 3001: ");
                ExitExtraction();
                return;
            }

            ExitExtraction(true);
        }

        private void ExitExtraction(bool success = false)
        {
            this.compressedFile.Processing = false;
            this.compressedFile.WasSuccessful = success;
        }
    }
}